Windows Python wheels: Bad address src\epoll.cpp:100

Hello, recently I set up a new python environment on a new computer as usual, I got automatically installed `pyzmq==21.0.2`, and I cannot use jupyter notebook, Jupyter lab, vs code at all, the python kernel crashes every time when I open a editor, and after one second, it directly crashes and lose connection to the kernel.

This problem annoyed me for several weeks, and today I saw many people encountered this issue From December 2020 to January 2021. And I have never encountered this issue in the past.

Also, this error message is somewhat not useful and not making sense, finally I realized its pyzmq error. 
I degraded it to `pyzmq==19.0.2` , it finally resolved this issue.

I hope this could be fixed, or if you could push old versions of pyzmq for users to install as dependencies. 
Thank you very much!
![IMG_1690](https://user-images.githubusercontent.com/36641890/107618797-1eed7880-6c8d-11eb-87ce-59d5acb13bf8.JPG)

I read all the posts here and on stack overflow and I wasn't able to find a solution that works for me.

I use Windows 10, Python 3.12.3 (installed via the app store). No VS Code or notebook whatever... I simply run my py script that communicates to my C++ code (running libzmq 4.3.5).

```
Bad file descriptor (C:\Users\runneradmin\AppData\Local\Temp\tmpkywtyyea\build\_deps\bundled_libzmq-src\src\epoll.cpp:73)
```

I tried to uninstall pyzmq and install an older version (tried different versions) but I wasn't able to do it. Every time I get the same error.

```
PS > pip install pyzmq==25.0.2
Defaulting to user installation because normal site-packages is not writeable
Collecting pyzmq==25.0.2
  Downloading pyzmq-25.0.2.tar.gz (1.2 MB)
     ---------------------------------------- 1.2/1.2 MB 3.7 MB/s eta 0:00:00
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
Building wheels for collected packages: pyzmq
  Building wheel for pyzmq (pyproject.toml) ... error
  error: subprocess-exited-with-error

  × Building wheel for pyzmq (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> [417 lines of output]
      C:\Users\ademyankov\AppData\Local\Temp\pip-build-env-e5nik6y2\overlay\Lib\site-packages\setuptools\_distutils\dist.py:266: UserWarning: Unknown distribution option: 'cffi_modules'
        warnings.warn(msg)
```

This is the end of the error logs:

```
      Generating code
      Finished generating code
      building 'zmq.backend.cython._device' extension
      creating build\temp.win-amd64-cpython-312\Release\zmq
      creating build\temp.win-amd64-cpython-312\Release\zmq\backend
      creating build\temp.win-amd64-cpython-312\Release\zmq\backend\cython
      "C:\Program Files\Microsoft Visual Studio\2022\Professional\VC\Tools\MSVC\14.39.33519\bin\HostX86\x64\cl.exe" /c /nologo /O2 /W3 /GL /DNDEBUG /MD -DWINVER=0x0602 -D_WIN32_WINNT=0x0602 -Ibundled\zeromq\include -Izmq\utils "-IC:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.12_3.12.1008.0_x64__qbz5n2kfra8p0\include" "-IC:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.12_3.12.1008.0_x64__qbz5n2kfra8p0\Include" "-IC:\Program Files\Microsoft Visual Studio\2022\Professional\VC\Tools\MSVC\14.39.33519\include" "-IC:\Program Files\Microsoft Visual Studio\2022\Professional\VC\Tools\MSVC\14.39.33519\ATLMFC\include" "-IC:\Program Files\Microsoft Visual Studio\2022\Professional\VC\Auxiliary\VS\include" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.22621.0\ucrt" "-IC:\Program Files (x86)\Windows Kits\10\\include\10.0.22621.0\\um" "-IC:\Program Files (x86)\Windows Kits\10\\include\10.0.22621.0\\shared" "-IC:\Program Files (x86)\Windows Kits\10\\include\10.0.22621.0\\winrt" "-IC:\Program Files (x86)\Windows Kits\10\\include\10.0.22621.0\\cppwinrt" /Tczmq\backend\cython\_device.c /Fobuild\temp.win-amd64-cpython-312\Release\zmq\backend\cython\_device.obj
      _device.c
      zmq\backend\cython\_device.c(4372): error C2039: 'curexc_traceback': is not a member of '_ts'
      C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.12_3.12.1008.0_x64__qbz5n2kfra8p0\include\cpython/pystate.h(115): note: see declaration of '_ts'
      zmq\backend\cython\_device.c(4375): error C2039: 'curexc_traceback': is not a member of '_ts'
      C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.12_3.12.1008.0_x64__qbz5n2kfra8p0\include\cpython/pystate.h(115): note: see declaration of '_ts'
      zmq\backend\cython\_device.c(5408): error C2039: 'ob_digit': is not a member of '_longobject'
      C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.12_3.12.1008.0_x64__qbz5n2kfra8p0\include\cpython/longintrepr.h(87): note: see declaration of '_longobject'
      zmq\backend\cython\_device.c(5463): error C2039: 'ob_digit': is not a member of '_longobject'
      C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.12_3.12.1008.0_x64__qbz5n2kfra8p0\include\cpython/longintrepr.h(87): note: see declaration of '_longobject'
      zmq\backend\cython\_device.c(5680): error C2039: 'ob_digit': is not a member of '_longobject'
      C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.12_3.12.1008.0_x64__qbz5n2kfra8p0\include\cpython/longintrepr.h(87): note: see declaration of '_longobject'
      zmq\backend\cython\_device.c(5735): error C2039: 'ob_digit': is not a member of '_longobject'
      C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.12_3.12.1008.0_x64__qbz5n2kfra8p0\include\cpython/longintrepr.h(87): note: see declaration of '_longobject'
      zmq\backend\cython\_device.c(6086): error C2039: 'ob_digit': is not a member of '_longobject'
      C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.12_3.12.1008.0_x64__qbz5n2kfra8p0\include\cpython/longintrepr.h(87): note: see declaration of '_longobject'
      DEBUG EXT_SUFFIX '.cp312-win_amd64' '.cp312-win_amd64.pyd'
      ************************************************
      DEBUG EXT_SUFFIX '.cp312-win_amd64' '.cp312-win_amd64.pyd'
      ************************************************
      error: command 'C:\\Program Files\\Microsoft Visual Studio\\2022\\Professional\\VC\\Tools\\MSVC\\14.39.33519\\bin\\HostX86\\x64\\cl.exe' failed with exit code 2
      [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
  ERROR: Failed building wheel for pyzmq
Failed to build pyzmq
ERROR: Could not build wheels for pyzmq, which is required to install pyproject.toml-based projects
```

All was working before with the older version of pyzmq. But now I can't even downgrade the version.

Any advice people?

--- a/docs/source/changelog.md
+++ b/docs/source/changelog.md
@@ -7,6 +7,16 @@ For a full changelog, consult the [git log](https://github.com/zeromq/pyzmq/comm
 
 ## 26
 
+### 26.2
+
+Re-disable IPC in Windows wheels.
+
+Installing pyzmq from source on Windows should build with IPC enabled.
+
+IPC support via epoll on Windows was disabled in pyzmq 24 due to crashes,
+this was reintroduced somewhat unintentionally via the new build system in 26.0,
+but unfortunately the crashes remain, so IPC is disabled again in 26.2.
+
 ### 26.1.1
 
 Windows wheels now statically link msvcp instead of bundling msvcp.dll, which could cause compatibility problems.

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -178,14 +178,16 @@ repair-wheel-command = """\
 """
 
 [tool.cibuildwheel.windows.config-settings]
+"cmake.define.ZMQ_PREFIX" = "bundled"
 # statically link MSVCP
 # see https://github.com/zeromq/pyzmq/issues/2012
 # and https://github.com/matplotlib/matplotlib/pull/28687
 "cmake.define.CMAKE_MSVC_RUNTIME_LIBRARY" = "MultiThreaded"
 "cmake.define.CMAKE_SHARED_LINKER_FLAGS" = "ucrt.lib;vcruntime.lib;/nodefaultlib:libucrt.lib;/nodefaultlib:libvcruntime.lib"
-
-[tool.cibuildwheel.windows.environment]
-ZMQ_PREFIX = "bundled"
+# disable IPC/epoll on Windows
+# due to https://github.com/zeromq/pyzmq/issues/1981
+"cmake.define.ZMQ_HAVE_IPC" = "OFF"
+"cmake.define.POLLER" = "select"
 
 # mac-arm target is 10.15
 [[tool.cibuildwheel.overrides]]

--- a/tools/test_wheel.py
+++ b/tools/test_wheel.py
@@ -20,7 +20,12 @@ except ImportError:
 def test_has(feature):
     import zmq
 
-    assert zmq.has(feature)
+    if feature == 'ipc' and sys.platform.startswith('win32'):
+        # IPC support is broken in enough cases on Windows
+        # that we can't ship wheels with it (for now)
+        assert not zmq.has(feature)
+    else:
+        assert zmq.has(feature)
 
 
 def test_simple_socket():

